substrate-node/pallets/pallet-tfgrid/src/farm.rs

```diff
@@ -1,10 +1,373 @@
+        let mut farm = Farms::<T>::get(farm_id).ok_or(Error::<T>::FarmNotExists)?;
+
+        let twin = Twins::<T>::get(farm.twin_id).ok_or(Error::<T>::TwinNotExists)?;
+        ensure!(
+            twin.account_id == account_id,
+            Error::<T>::CannotUpdateFarmWrongTwin
+        );
+
+        // Check if it's a valid IP4
+        let ip4 = IP4 { ip, gw };
+        ip4.is_valid().map_err(|_| Error::<T>::InvalidPublicIP)?;
+
+        let new_ip = PublicIP {
+            ip: ip4.ip,
+            gateway: ip4.gw,
+            contract_id: 0,
+        };
+
+        match farm
+            .public_ips
+            .iter()
+            .position(|public_ip| public_ip.ip == new_ip.ip)
+        {
+            Some(_) => return Err(Error::<T>::IpExists.into()),
+            None => {
+                farm.public_ips
+                    .try_push(new_ip)
+                    .map_err(|_| Error::<T>::InvalidPublicIP)?;
+                Farms::<T>::insert(farm.id, &farm);
+                Self::deposit_event(Event::FarmUpdated(farm));
+                return Ok(().into());
+            }
+        };
+    }
+
+    pub fn _remove_farm_ip(
+        account_id: T::AccountId,
+        id: u32,
+        ip: Ip4Input,
+    ) -> DispatchResultWithPostInfo {
+        let mut farm = Farms::<T>::get(id).ok_or(Error::<T>::FarmNotExists)?;
+
+        let twin = Twins::<T>::get(farm.twin_id).ok_or(Error::<T>::TwinNotExists)?;
+        ensure!(
+            twin.account_id == account_id,
+            Error::<T>::CannotUpdateFarmWrongTwin
+        );
+
+        match farm
+            .public_ips
+            .iter()
+            .position(|pubip| pubip.ip == ip && pubip.contract_id == 0)
+        {
+            Some(index) => {
+                farm.public_ips.remove(index);
+                Farms::<T>::insert(farm.id, &farm);
+                Self::deposit_event(Event::FarmUpdated(farm));
+                Ok(().into())
+            }
+            None => Err(Error::<T>::IpNotExists.into()),
+        }
+    }
+
+    pub fn _delete_node_farm(account_id: T::AccountId, node_id: u32) -> DispatchResultWithPostInfo {
+        // check if the farmer twin is authorized
+        let farm_twin_id =
+            TwinIdByAccountID::<T>::get(&account_id).ok_or(Error::<T>::TwinNotExists)?;
+        // check if the ndode belong to said farm
+        let node = Nodes::<T>::get(&node_id).ok_or(Error::<T>::NodeNotExists)?;
+        let farm = Farms::<T>::get(node.farm_id).ok_or(Error::<T>::FarmNotExists)?;
+
+        ensure!(
+            Twins::<T>::contains_key(&farm.twin_id),
+            Error::<T>::TwinNotExists
+        );
+        let farm_twin = Twins::<T>::get(farm.twin_id).ok_or(Error::<T>::TwinNotExists)?;
+        ensure!(
+            farm_twin_id == farm_twin.id,
+            Error::<T>::FarmerNotAuthorized
+        );
+
+        let mut nodes_by_farm = NodesByFarmID::<T>::get(node.farm_id);
+        let location = nodes_by_farm
```

Comment:
Indeed at this stage we know if logic is respected that the node must be in the of farm nodes list. 
I think the point here is to get the position of the node in the list.
And since it s a search that returns a result it needs to handle the "nothing found" situation just in case the search is done the wrong way for example.

Reply:
Yes but if the node id is not found in the list here it will return. So the node will not be deleted while that is the intention of the function. And if the node id is not in the list that's what we want in the end (after the function). So that's what I'm wondering if we should really return an error here.

Reply:
Ok, I see your point
Let s unblock the removal in that case